kerastuner/applications/augment.py

```diff
@@ -0,0 +1,191 @@
+            transform_layer = TRANSFORMS[transform](transform_factor)
+            model.add(transform_layer)
+
+        return model
+
+class HyperRandAugment(HyperAugment):
+    """An HyperModel for Rand augmentation.
+       Based on https://arxiv.org/abs/1909.13719.
+       This augmentation randomly picks `randaug_count` augmentation transforms
+       from a pool of transforms for each sample, and set the augmentation
+       magnitude to `randaug_mag`.
+       Only supporting augmentations in Keras preprocessing layers.
+
+    # Arguments:
+        input_shape: Optional shape tuple, e.g. `(256, 256, 3)`.
+              One of `input_shape` or `input_tensor` must be
+              specified.
+        input_tensor: Optional Keras tensor (i.e. output of
+            `layers.Input()`) to use as image input for the model.
+              One of `input_shape` or `input_tensor` must be
+              specified.
+        rotate: Whether to include rotation in augmentation. When
+             set to True, RandomRotation layer will be added to the
+             pool of augmentations to select from.
+        translate_x: Whether to include horizontal translation in augmentation.
+        translate_y: Whether to include vertical translation in augmentation.
+        contrast: Whether to include random contrast in augmentation.
+        **kwargs: Additional keyword arguments that apply to all
+            HyperModels. See `kerastuner.HyperModel`.
+
+    # Search space:
+        'randaug_mag': Float, default range [0.05, 1].
+                       Controls the strength of each augmentation
+                       applied to each image.
+        'randaug_count': Int, default range [0, 5].
+                       Controls the number of augmentations layers
+                       applied to each image.
+    """
+    def __init__(self,
+                 input_shape=None,
+                 input_tensor=None,
+                 **kwargs):
+        super(HyperRandAugment, self).__init__(input_shape=input_shape,
+            input_tensor=input_tensor,
+            **kwargs)
+        if input_shape is None and input_tensor is None:
+            raise ValueError('You must specify either `input_shape` or '
+                             '`intput_tensor` when using HyperRandAugment.')
+
+    def build(self, hp):
+        if self.input_tensor is not None:
+            inputs = keras.utils.get_source_inputs(self.input_tensor)
+            x = self.input_tensor
+        else:
+            inputs = layers.Input(shape=self.input_shape)
+            x = inputs
+
+        randaug_mag = hp.Float('randaug_mag', 0.05, 1.0, step=0.1, default=0.2)
+        randaug_count = hp.Int('randaug_count', 0, 5, default=2)
+
+        for _ in range(randaug_count):
+            # selection tensor determines for each sample, which operation
+            # is used.
+            batch_size = tf.shape(x)[0]
+            selection = tf.random.uniform([batch_size, 1, 1, 1],
```

Comment:
I am bit confused about why this need to be a 4D tensor.

Reply:
RandAugment selects a random operation on each sample in a batch (This is not very clearly and directly stated in the paper, but can be inferred from their [code](https://github.com/tensorflow/tpu/blob/master/models/official/efficientnet/autoaugment.py)), so this random number `selection` should be generated on the batch dimension.